lib/domain.js

```diff
@@ -387,3 +382,49 @@ Domain.prototype.bind = function(cb) {
 
   return runBound;
 };
+
+// Override EventEmitter methods to make it domain-aware.
+EventEmitter.usingDomains = true;
+
+const eventInit = EventEmitter.init;
+EventEmitter.init = function() {
+  this.domain = null;
+  if (exports.active && !(this instanceof exports.Domain)) {
+    this.domain = exports.active;
+  }
+
+  return eventInit.call(this);
+};
+
+const eventEmit = EventEmitter.prototype.emit;
+EventEmitter.prototype.emit = function emit(...args) {
```

Comment:
@vdeturckheim Would you mind changing this to `emit()` and then using `arguments` everywhere? I realize performance is not that big of a deal but this will actually make a substantial difference.

Reply:
@apapirovski are you sure it's faster?

Reply:
Yeah, I'm certain because I checked it very recently. `...args` is only faster when arguments requires slicing or manually copying in situations like `emit(type, ...args)`.

Reply:
@apapirovski very interesting can you share a benchmark showing that? I'm pretty sure `...args` is optimized. 

cc @bmeurer

Reply:
It has been optimized but in practical usage it still seems to perform worse. I've run into this working on `nextTick`, `events` and `timers`.

```js
> function fn(...args) {
...   return Reflect.apply(Math.max, Math, args);
... }
> 
> function fn2() {
...   return Reflect.apply(Math.max, Math, arguments);
... }
> console.time('test'); for (var i = 0; i < 1e8; i++) { fn(1, 2, 3, 4); } console.timeEnd('test');
test: 279.502ms
> console.time('test'); for (var i = 0; i < 1e8; i++) { fn2(1, 2, 3, 4); } console.timeEnd('test');
test: 284.195ms
```

The basic benchmark shows it being optimized. As soon as I've tried using it in practice where there's more than one layer to the usage, arguments seems to win out.

Reply:
btw @benjamingr I decided to finally split test all the possible variations. Here are my findings:

```
function fn(...args) {
  cb(...args);
}
```

is the fastest for that particular purpose, faster than using `arguments`. But then there's:

```
function fn(something, ...args) {
  cb(...args);
}
```

seems to depend on specific usage of `args` within `fn`. For example, I've found that in `setTimeout` and `setImmediate` we incur about 20-25% performance penalty compared to using manual copying of `arguments`.

On the flipside, it performs as expected within `EventEmitter.prototype.emit`... there definitely seem to be some edge cases around this behaviour that I don't quite understand.

I don't know if this helps with your plan to bug V8 devs regarding this :)

Reply:
It's possible this might be related to whether `args` is being stored somewhere or just passed through to another function, or whether `args` is accessed at all in the body?